### DISCUSSION TITLE
I find spans confusing

### DISCUSSION THREAD

**Discussion by Kixunil** [nushell/nushell#7018] [General]
When digging into the code I found spans to be quite confusing. There is multiple of them and it's not clear which is meant to point at what and it seems sometimes it'd be good to have multiple spans that turn into nice error messages. There seems to be an attempt to do something like this in json convertor but the resulting error messages are not as good as I'd expect.

Somewhat ironically, the enum fields containing spans are often unnamed or jut named `span` while e.g. `Value::Error` contains a single obvious field that is named forcing less elegant code like `unwrap_or_else(|error| Value::Error { error })` that could've been `unwrap_or_else(Value::Error)` instead. (Disclaimer: I do not intend to disrespect the authors, I totally understand that code is often not 100% perfect for various reasons. I actually appreciate that there's something working available. Nu is great!)

## Examples

Note that the suggested error messages only intend to convey the concept of referring to multiple things clearly.
They are NOT suggestions about the layout of the messages.

### Direct input

Command: `"{foo}" | from json`

Potentially interesting spans: data span pointing at `{foo}` and command span pointing at `from json`

Potential nice error message:

```
Can not parse json: unexpected token

"{foo}" | from json
          ^^^^^^^^^^ - this command failed

The input to the command `from json` is invalid:

{foo}
 ^^^ - unexpected token

Expected a json dictionary key starting with `"`

"{foo}" | from json
^^^^^^^ - the input originates here
```

The current error message:

```
Error: 
  × Error while parsing JSON text
   ╭─[entry #392:1:1]
 1 │ "{foo}" | from json
   ·           ────┬────
   ·               ╰── error parsing JSON text
   ╰────

Error: 
  × Error while parsing JSON text
   ╭────
 1 │ {foo}
   ╰────

```

Notice that the error message may look like `from json` is an invalid json text - but it is a command.
Also note that the exact location of the error potentially could've been marked but it isn't.

### External program

Command: `some_external_program | from json` (assume it produces `{foo}` again)

Potentially interesting spans:origin span pointing at `some_external_program` and command span pointing at `from json`

Potential nice error message:

```
Can not parse json: unexpected token

some_external_program | from json
                        ^^^^^^^^^^ - this command failed

The input to the command `from json` is invalid:

{foo}
 ^^^ - unexpected token

Expected a json dictionary key starting with `"`

some_external_program | from json
^^^^^^^^^^^^^^^^^^^^^ - the input was produced by this command
```

The current error message:

```
Error: 
  × Error while parsing JSON text
   ╭─[entry #395:1:1]
 1 │ some_external_program | from json
   ·                         ────┬────
   ·                                   ╰── error parsing JSON text
   ╰────

Error: 
  × Error while parsing JSON text
   ╭────
 1 │ {foo}
   ╰────

```

There's:

> // TODO: turn this into a structured underline of the nu_json error

Suggesting that maybe what I propose is possible but not implemented. I might try to fix it but the confusion around spans doesn't help and the comment doesn't contain any hint either. :(

**Comment by rgwood** [comment]
Agreed that we could use some clean up and maybe additional tooling in this area.

Something that I find difficult: when running Nu in a debugger, I haven't found a good way to see which text a `Span` refers to.
